### ISSUE TITLE
Press F5 to run code and get BufferExceptionTypeError then can not run the code successfully

### ISSUE THREAD

**Issue by FIRECONTER** [foreshadow/atom-python-run#67]
I use the latest version. But I meet such error:

 atom-python-run: BufferExceptionTypeError: Cannot read property 'then' of undefined: You failed to create the file before attempting to execute it. You must create the file first and then execute it after it has been created.

I have checked that the error comes from the function saveAndRun in atom-python-run.js file

atom.workspace.getActiveTextEditor().save().then(() => {
                    run({
                        'command': atom.config.get(`atom-python-run.${key}Command`),
                        'pause': atom.config.get(`atom-python-run.${key}Pause`),
                    });
                });

**Comment by FIRECONTER** [comment]
if I choose an earlier version such as V0.91, it works well

**Comment by xovertheyearsx** [comment]
Please describe steps to reproduce this error in detail. Step by step from beginning to end, how do you trigger the error?

**Comment by neederhow** [comment]
I meet this error too.
I just open a .py file and press f5\f6 and the error message popup

atom-python-run: BufferExceptionTypeError: Cannot read property 'then' of undefined: You failed to create the file before attempting to execute it. You must create the file first and then execute it after it has been created.

**Comment by xovertheyearsx** [comment]
For any users/developers experiencing similar issues posting a comment; please consider the following questions:

- did you read the [README.md](https://github.com/foreshadow/atom-python-run/wiki/00-Readme)?
- what version of `atom` are you using?
- what version of `atom-python-run` are you using?
- are you using `atom-python-run`s default settings?
- are you using `atom`s [console]() to help you debug?
- did the `.py` file previously exist? (did you `save as` first and then `f5` after?)
- does this happen with **any** `.py` files? (where files previously existed and failed to execute?)
- can you execute your `.py` file using `python` or `idle` without errors?
- do the file paths use spaces? (i.e. `c:\users\user name\documents\my python\hello world.py`)
  - if the file paths use spaces, refer to issue #60 instead, and note that issue #67 led you there

To verify what @FIRECONTER originally stated, this [block](https://github.com/foreshadow/atom-python-run/blob/master/lib/atom-python-run.js#L19) causes this issue.

```javascript
atom.workspace.getActiveTextEditor().save().then(() => {
    run({
        'command': atom.config.get(`atom-python-run.${key}Command`),
        'pause': atom.config.get(`atom-python-run.${key}Pause`),
    });
});
```

From what I can tell, this is because [save().then()](https://atom.io/docs/api/v1.23.1/TextBuffer#instance-save) acts as a [Promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) object which [expects something to be returned](https://stackoverflow.com/questions/24788171/typeerror-cannot-read-property-then-of-undefined#24788320). 

The error `TypeError: Cannot read property 'then' of undefined` reflects that it expects the function `saveAndRun()` to return something signifying that the promise has been resolved. 

Because nothing is returned, the [try...catch](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Control_flow_and_error_handling#try...catch_statement) statement catches this exception. The exception is mangled by the text that is added to error using `atom.notifications.addError()`.

For a whole picture, this is what it looks like now.
```javascript
function saveAndRun(key) {
    try {
        console.log(`${key} pressed`)
        atom.workspace.getActiveTextEditor().save().then(() => {
            run({
                'command': atom.config.get(`atom-python-run.${key}Command`),
                'pause': atom.config.get(`atom-python-run.${key}Pause`),
            });
        });
    } catch(error) {
        atom.notifications.addError(`atom-python-run: \`BufferException${error}\`: You failed to create the file before attempting to execute it. You must create the file first and then execute it after it has been created.`);
        return;
    }
}
```

In theory, since we expect a `Promise` to be returned, we could just say `return save.then(run())` and clean up the error message. 

Because I don't know exactly what causes the issue, I can't be sure. I haven't been able to reproduce this error at all. I need in depth, step by step, information so I can reproduce it. Although, if someone figured that out, I'm sure they would be able to surmise a fix for it. 

For now, It could be changed to this instead.

```javascript
function saveAndRun(key) {
    try {
        console.log(`${key} pressed`);
        // save() returns a promise
        let save = atom.workspace.getActiveTextEditor().save();
        // Atom expects that promise to be resolved or rejected
        return save.then(() => {
            run({
                'command': atom.config.get(`atom-python-run.${key}Command`),
                'pause': atom.config.get(`atom-python-run.${key}Pause`),
            });
        });
    } catch(error) {
        // the promise failed and the exception was caught
        console.log(`atom-python-run: saveAndRun: ${error}`);
        atom.notifications.addError(`atom-python-run: \`${error}\`.`);
        // nothing to do, so we return undefined here
        return;
    }
}
```

Again, I can't guarantee that this will work. I would need someone to test this, see if it works, and then report the results here.

**Comment by xovertheyearsx** [comment]
Look [here](https://github.com/foreshadow/atom-python-run/wiki/12-How-Do-I-Use-atom-python-run#console-log) for help on how to access and use the Console Log. 

You can see [here](https://github.com/foreshadow/atom-python-run/wiki/12-How-Do-I-Use-atom-python-run#console-vs-cp-log) that the Console Log and cp Log are different logging mechanisms.

Feel free to express if you're confused and/or need additional clarification. If you've found a fix, please post it here, make a pull request, or create a fork with an implemented fix.

**Comment by foreshadow** [comment]
Fixed in v0.9.6
